build: Auto-load ICU data from --with-icu-default-data-dir

sgallagher opened the pull request:
When compiled with `--with-intl=small` and
`--with-icu-default-data-dir=PATH`, Node.js will use PATH as a
fallback location for the ICU data.

We will first perform an access check using fopen(PATH, 'r') to
ensure that the file is readable. If it is, we'll set the
icu_data_directory and proceed. There's a slight overhead for the
fopen() check, but it should be barely measurable.

This will be useful for Linux distribution packagers who want to
be able to ship a minimal node binary in a container image but
also be able to add on the full i18n support where needed. With 
this patch, it becomes possible to ship the interpreter as 
/usr/bin/node in one package for the distribution and to ship the
data files in another package (without a strict dependency 
between the two). This means that users of the distribution will
not need to explicitly direct Node.js to locate the ICU data. It
also means that in environments where full internationalization is
not required, they do not need to carry the extra content (with 
the associated storage costs).


Related: https://github.com/nodejs/node/issues/3460

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

Attn: @srl295 

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [X] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [X] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

srl295 commented on src/node.cc:
```diff
@@ -848,6 +849,26 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
+
+    if (f) {
+      fclose(f);
+      per_process::cli_options->icu_data_dir = NODE_ICU_DEFAULT_DATA_DIR;
```
^ Leverage… nice.

bnoordhuis commented on configure.py:
```diff
@@ -1373,6 +1381,11 @@ def write_config(data, name):
     o['variables']['icu_locales'] = ','.join(str(loc) for loc in locs)
     # We will check a bit later if we can use the canned deps/icu-small
+    o['variables']['icu_has_default_data_dir'] = b(False)
+    if bool(options.with_icu_default_data_dir):
```
```suggestion
    if options.with_icu_default_data_dir:
```

bnoordhuis commented on configure.py:
```diff
@@ -1373,6 +1381,11 @@ def write_config(data, name):
+    o['variables']['icu_has_default_data_dir'] = b(False)
+    if bool(options.with_icu_default_data_dir):
+      o['variables']['icu_has_default_data_dir'] = b(True)
+      o['variables']['icu_default_data_dir'] = options.with_icu_default_data_dir      # Needs to be set but empty for node.gypi to process
```
Long line. FWIW, I'd just set it like this:
```python
o['variables']['icu_default_data_dir'] = options.with_icu_default_data_dir or ''
```
(e.g., no if stmt, no `icu_has_default_data_dir`) and then just check that `icu_default_data_dir!=""` in the gyp files.

sgallagher replied:
I actually tried exactly this first, but for some reason the `icu_default_data_dir!=''` was always failing the comparison. I'm not sure why. I'll try again.

sgallagher replied:
Seems to be working in that construct, so :man_shrugging:

sgallagher replied:
Nope, I see what it was now. It appears that gyp treats variables suffixed with `_dir` specially and it was throwing:

```
Traceback (most recent call last):
  File "./configure", line 28, in <module>
    import configure
  File "/home/sgallagh/workspace/node/configure.py", line 1729, in <module>
    run_gyp(gyp_args)
  File "tools/gyp_node.py", line 54, in run_gyp
    rc = gyp.main(args)
  File "tools/gyp/pylib/gyp/__init__.py", line 547, in main
    return gyp_main(args)
  File "tools/gyp/pylib/gyp/__init__.py", line 523, in gyp_main
    options.duplicate_basename_check)
  File "tools/gyp/pylib/gyp/__init__.py", line 139, in Load
    params['parallel'], params['root_targets'])
  File "tools/gyp/pylib/gyp/input.py", line 2785, in Load
    variables, includes, depth, check, True)
  File "tools/gyp/pylib/gyp/input.py", line 383, in LoadTargetBuildFile
    includes, True, check)
  File "tools/gyp/pylib/gyp/input.py", line 255, in LoadOneBuildFile
    aux_data, includes, check)
  File "tools/gyp/pylib/gyp/input.py", line 293, in LoadBuildFileIncludesIntoDict
    subdict_path, include)
  File "tools/gyp/pylib/gyp/input.py", line 2151, in MergeDicts
    MergeDicts(to[k], v, to_file, fro_file)
  File "tools/gyp/pylib/gyp/input.py", line 2144, in MergeDicts
    to[k] = MakePathRelative(to_file, fro_file, v)
  File "tools/gyp/pylib/gyp/input.py", line 2043, in MakePathRelative
    if item[-1] == '/':
IndexError: string index out of range while reading includes of /home/sgallagh/workspace/node/node.gyp while trying to load /home/sgallagh/workspace/node/node.gyp
```

After I drop the `_dir` suffix (leaving `icu_default_data`), it works properly.

bnoordhuis commented on src/node.cc:
```diff
@@ -848,6 +849,26 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
     credentials::SafeGetenv("NODE_ICU_DATA",
                             &per_process::cli_options->icu_data_dir);
+
+  #ifdef NODE_ICU_DEFAULT_DATA_DIR
```
```suggestion
#ifdef NODE_ICU_DEFAULT_DATA_DIR
```
(don't indent preprocessor directives)

bnoordhuis commented on src/node.cc:
```diff
@@ -848,6 +849,26 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
+    char full_path[2048];
+
+    strcpy(full_path,
+           NODE_ICU_DEFAULT_DATA_DIR U_FILE_SEP_STRING NODE_ICU_DATA_FILE);
```
Don't use strcpy() here since it's a static string, do this instead:
```c++
static const char full_path[] = NODE_ICU_DEFAULT_DATA_DIR "/" NODE_ICU_DATA_FILE;
```
(I don't think `U_FILE_SEP_STRING` is necessary since `/` is always a valid path separator on all platforms we support.)

bnoordhuis commented on src/node.cc:
```diff
@@ -848,6 +849,26 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
+
+    strcpy(full_path,
+           NODE_ICU_DEFAULT_DATA_DIR U_FILE_SEP_STRING NODE_ICU_DATA_FILE);
+    FILE *f = fopen(full_path, "r");
```
```suggestion
    FILE* f = fopen(full_path, "rb");
```

bnoordhuis commented on src/node.cc:
```diff
@@ -848,6 +849,26 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
+           NODE_ICU_DEFAULT_DATA_DIR U_FILE_SEP_STRING NODE_ICU_DATA_FILE);
+    FILE *f = fopen(full_path, "r");
+
+    if (f) {
```
```suggestion
    if (f != nullptr) {
```

sgallagher commented:
I updated the patch with the code review recommendations. I also changed the commit message and the first comment on this PR to match it and provide more detail on the problem this is solving.

Thank you very much for the review!

richardlau commented:
Shouldn't this target master?

sgallagher commented:
Rebased it to master. I was working on it against v12 originally because a) that's what we have in Fedora and b) the master branch wasn't compiling successfully for me, but that's been fixed in the interim it seems.

srl295 commented:
fyi @nodejs/intl

bnoordhuis commented on src/node.cc:
```diff
@@ -90,6 +90,7 @@
 
 #if defined(NODE_HAVE_I18N_SUPPORT)
 #include <unicode/uvernum.h>
+#include <unicode/putil.h>
```
```suggestion
#include <unicode/utypes.h>
```
(I think that's the right header when you only need `U_ICUDATA_NAME`.)

bnoordhuis commented on src/node.cc:
```diff
@@ -872,6 +873,25 @@ int InitializeNodeWithArgs(std::vector<std::string>* argv,
+    // Check whether the NODE_ICU_DEFAULT_DATA_DIR contains the right data
+    // file and can be read.
+    static const char full_path[] =
+      NODE_ICU_DEFAULT_DATA_DIR "/" U_ICUDATA_NAME ".dat";
```
```suggestion
        NODE_ICU_DEFAULT_DATA_DIR "/" U_ICUDATA_NAME ".dat";
```

sgallagher commented:
Rebased atop the latest master and made the two minor changes @bnoordhuis requested.

Trott commented:
Landed in d502b83
